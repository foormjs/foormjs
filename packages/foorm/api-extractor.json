{
  "extends": "../../api-extractor.json",
  "mainEntryPointFilePath": "./dist/packages/foorm/src/index.d.ts",
  "dtsRollup": {
    "publicTrimmedFilePath": "./dist/index.d.ts"
  },
  "compiler": {
    "overrideTsconfig": {
      "compilerOptions": {
        "paths": {
          "common": ["packages/foorm/dist/common"]
        }
      },         
      "include": [
        "packages/foorm/dist/packages",
        "packages/foorm/dist/common"
      ]
    }
  }
}