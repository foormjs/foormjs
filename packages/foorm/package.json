{
  "name": "foorm",
  "version": "0.0.2-alpha.5",
  "description": "foorm",
  "main": "dist/index.cjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    }
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/foormjs/foormjs.git",
    "directory": "packages/foorm"
  },
  "keywords": [
    "foorm",
    "foormjs",
    "forms",
    "validations"
  ],
  "author": "Artem Maltsev",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/foormjs/foormjs/issues"
  },
  "homepage": "https://github.com/foormjs/foormjs/tree/main/packages/foorm#readme",
  "dependencies": {
    "@prostojs/ftring": "^0.0.3"
  }
}
