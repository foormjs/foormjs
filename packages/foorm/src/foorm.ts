/* eslint-disable no-inner-declarations */
/* eslint-disable @typescript-eslint/naming-convention */
/* eslint-disable @typescript-eslint/no-explicit-any */

import type {
  TComputed,
  TFoormEntry,
  TFoormEntryOptions,
  TFoormFnScope,
  TFoormMetaExecutable,
} from './types'
import { evalParameter } from './utils'

export interface TFoormOptions<D, C> {
  title?: TComputed<string, D, C>
  entries: Array<TFoormEntry<any, D, C, TFoormEntryOptions>>
  submit?: TFoormMetaExecutable<D, C>['submit']
  context?: C
}

export class Foorm<D = any, C = any> {
  protected entries: Array<TFoormEntry<any, D, C, TFoormEntryOptions>>

  protected submit?: TFoormMetaExecutable<D, C>['submit']

  protected title?: TComputed<string, D, C>

  protected context: C

  // private fns!: FNPool<string | boolean, TFoormFnScope>

  constructor(opts?: TFoormOptions<D, C>) {
    this.entries = opts?.entries || []
    this.submit = opts?.submit
    this.title = opts?.title || ''
    this.context = opts?.context || ({} as C)
  }

  public addEntry<V>(entry: TFoormEntry<V, D, C, TFoormEntryOptions>) {
    this.entries.push(entry as TFoormEntry<unknown, D, C, TFoormEntryOptions>)
  }

  public setTitle(title: string) {
    this.title = title
  }

  public setSubmit(submit: TFoormMetaExecutable<D, C>['submit']) {
    this.submit = submit
  }

  public setContext(context: C) {
    this.context = context
  }

  getDefinition() {
    return {
      title: this.title,
      submit: this.submit || ({ text: 'Submit' } as TFoormMetaExecutable<D, C>['submit']),
      context: this.context,
      entries: this.entries,
    }
  }

  protected normalizeEntry<V, O extends TFoormEntryOptions>(
    e: TFoormEntry<V, D, C, O>
  ): TFoormEntry<V, D, C, O> & {
    name: string
    label: TComputed<string, D, C>
    type: string
  } {
    return {
      ...e,
      name: e.name || e.field,
      label: (e.label || e.field) as string,
      type: e.type || 'text',
    }
  }

  public executable(): TFoormMetaExecutable<D, C> {
    return {
      title: this.title || '',
      submit: this.submit || ({ text: 'Submit' } as TFoormMetaExecutable<D, C>['submit']),
      context: this.context,
      entries: this.entries.map(e => this.normalizeEntry(e)),
    }
  }

  createFormData<T extends Record<string, unknown>>(): T {
    const data: T = {} as T
    for (const entry of this.entries) {
      if (entry.type !== 'action') {
        data[entry.field as keyof T] = (entry.value || undefined) as T[keyof T]
      }
    }
    return data
  }

  prepareValidators(_validators: TFoormEntry<any, D, C, TFoormEntryOptions>['validators']) {
    const validators =
      _validators || ([] as Required<TFoormEntry<any, D, C, TFoormEntryOptions>>['validators'])
    validators.unshift((v, _d, _c, entry) => entry.optional || !!v || 'Required')
    return validators
  }

  supportsAltAction(altAction: string) {
    return !!this.entries.some(e => e.altAction === altAction)
  }

  getFormValidator(): (inputs: Record<string, unknown>) => {
    passed: boolean
    errors: Record<string, string>
  } {
    const entries = this.executable().entries
    const fields: Record<
      string,
      {
        entry: TFoormEntry<any, D, C, TFoormEntryOptions>
        validators: TFoormEntry<any, D, C, TFoormEntryOptions>['validators']
      }
    > = {}
    for (const entry of entries) {
      if (entry.field) {
        const validators = this.prepareValidators(entry.validators)
        fields[entry.field] = {
          entry,
          validators,
        }
        fields[entry.field].validators = validators
      }
    }
    return (data: Record<string, unknown>) => {
      let passed = true
      const errors: Record<string, string> = {}
      for (const [key, value] of Object.entries(fields)) {
        const evalEntry = { ...value.entry }
        const scope: TFoormFnScope<any, D, C> = {
          v: data[key],
          context: this.context,
          entry: {
            field: evalEntry.field,
            type: evalEntry.type!,
            component: evalEntry.component,
            name: evalEntry.name!,
            length: evalEntry.length,
          },
          data: data as D,
        }

        if (scope.entry) {
          scope.entry.disabled = evalParameter<boolean>(evalEntry.disabled as boolean, scope, true)
          scope.entry.optional = evalParameter<boolean>(evalEntry.optional as boolean, scope, true)
          scope.entry.hidden = evalParameter<boolean>(evalEntry.hidden as boolean, scope, true)
        }
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const result = validate<any, D, C>({
          v: data[key],
          context: this.context,
          validators: value.validators,
          entry: scope.entry,
          data: data as D,
        })
        if (!result.passed) {
          passed = false
          if (!errors[key]) {
            errors[key] = result.error || 'Wrong value'
          }
        }
      }
      return {
        passed,
        errors,
      }
    }
  }
}

export function validate<V, D, C>(
  opts: TFoormFnScope<V, D, C> & {
    validators: TFoormEntry<any, D, C, TFoormEntryOptions>['validators']
  }
) {
  for (const validator of opts.validators || []) {
    const result = evalParameter(
      validator,
      {
        v: opts.v,
        context: opts.context,
        data: opts.data,
        entry: opts.entry,
      },
      true
    )
    if (result !== true) {
      return {
        passed: false,
        error: result || 'Wrong value',
      }
    }
  }
  return { passed: true }
}
