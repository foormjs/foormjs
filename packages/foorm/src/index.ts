export * from './deserialize'
export * from './foorm'
export * from './serialize'
export * from './types'
export * from './utils'
