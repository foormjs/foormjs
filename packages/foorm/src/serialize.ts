/* eslint-disable @typescript-eslint/naming-convention */
import { isCleanFn, serializeFn } from '@prostojs/serialize-fn'

import type { Foorm } from './foorm'
import type { TComputed, TComputedWithVal, TFoormSerialized, TSerializedFn } from './types'

export function serializeForm<D, C>(
  form: Foorm<D, C>,
  opts?: {
    replaceContext?: C
    replaceValues?: Record<string, unknown>
  }
): TFoormSerialized {
  const def = form.getDefinition()
  return {
    title: serializeComputedFn(def.title),
    submit: {
      text: serializeComputedFn(def.submit.text || 'Submit'),
      disabled: serializeComputedFn(def.submit.disabled) as boolean,
    },
    context: opts?.replaceContext || def.context,
    entries: def.entries.map(e => ({
      ...e,
      // strings
      label: serializeComputedFnWithVal(e.label) as string,
      description: serializeComputedFnWithVal(e.description) as string,
      hint: serializeComputedFnWithVal(e.hint) as string,
      placeholder: serializeComputedFnWithVal(e.placeholder) as string,
      // strings || objects
      classes: serializeComputedFnWithVal(
        e.classes as TComputedWithVal<string, any, D, C>,
        true
      ) as string,
      styles: serializeComputedFnWithVal(e.styles, true) as string,
      // booleans
      optional: serializeComputedFnWithVal(e.optional) as TSerializedFn,
      disabled: serializeComputedFnWithVal(e.disabled) as TSerializedFn,
      hidden: serializeComputedFnWithVal(e.hidden) as TSerializedFn,
      // options
      // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment
      options: serializeComputedFnWithVal(e.options) as any,
      // attrs
      attrs: serializeComputedFnWithVal(e.attrs, true) as Record<string, TSerializedFn>,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      value: opts?.replaceValues ? opts.replaceValues[e.field] : e.value,
      validators: (e.validators || [])
        .filter(v => isCleanFn(v))
        .map(fn => serializeComputedFnWithVal(fn)) as TSerializedFn[],
    })),
  }
}

function serializeComputedFn<OF, D, C>(fn?: TComputed<OF, D, C>): OF | TSerializedFn {
  if (typeof fn === 'function') {
    return { __fn__: serializeFn(fn as () => any, 'data', 'context') }
  }
  return fn as OF
}

function serializeComputedFnWithVal<OF, V, D, C>(
  fn?: TComputedWithVal<OF, V, D, C> | Record<string, TComputedWithVal<OF, V, D, C>>,
  inObj?: boolean
): Record<string, TSerializedFn> | OF | TSerializedFn {
  if (inObj && typeof fn === 'object') {
    const o = {} as Record<string, TSerializedFn>
    for (const [key, val] of Object.entries(fn as Record<string, TComputedWithVal<OF, V, D, C>>)) {
      o[key] = serializeComputedFnWithVal(val) as TSerializedFn
    }
    return o as OF | TSerializedFn
  }
  if (typeof fn === 'function') {
    return { __fn__: serializeFn(fn as () => any, 'v', 'data', 'context', 'entry') }
  }
  return fn as OF
}
