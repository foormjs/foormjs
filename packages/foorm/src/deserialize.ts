/* eslint-disable @typescript-eslint/naming-convention */
import { FNPool } from '@prostojs/deserialize-fn'

import { Foorm } from './foorm'
import type {
  TFoormEntryOptions,
  TFoormMetaExecutable,
  TFoormSerialized,
  TFoormValidatorFn,
  TSerializedFn,
} from './types'

const pool = new FNPool()

export function deserializeForm<D = unknown, C = unknown>(form: TFoormSerialized): Foorm<D, C> {
  return new Foorm({
    title: deserializeComputedFn(form.title) as string,
    context: form.context as C,
    submit: deserializeComputedFn(form.submit, true) as TFoormMetaExecutable<D, C>['submit'],
    entries: form.entries.map(e => ({
      ...e,
      // strings
      label: deserializeComputedFn(e.label) as string,
      description: deserializeComputedFn(e.description) as string,
      hint: deserializeComputedFn(e.hint) as string,
      placeholder: deserializeComputedFn(e.placeholder) as string,
      // strings || objects
      classes: deserializeComputedFn(e.classes, true) as string,
      styles: deserializeComputedFn(e.styles, true) as string,
      // booleans
      optional: deserializeComputedFn<boolean, unknown, unknown>(e.optional) as boolean,
      disabled: deserializeComputedFn<boolean, unknown, unknown>(e.disabled) as boolean,
      hidden: deserializeComputedFn<boolean, unknown, unknown>(e.hidden) as boolean,
      // options
      // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment
      options: deserializeComputedFn(e.options) as TFoormEntryOptions[],
      // attrs
      attrs: deserializeComputedFn(e.attrs, true) as Record<string, any>,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      value: e.value,
      validators: (e.validators || []).map(fn => deserializeComputedFn(fn)) as unknown as Array<
        TFoormValidatorFn<any, D, C>
      >,
    })),
  })
}

function deserializeComputedFn<T, D, C>(
  v?: TSerializedFn | T,
  asObj?: boolean
): T | ((__ctx__: unknown) => T) {
  if (typeof v === 'object' && typeof (v as TSerializedFn).__fn__ === 'string') {
    const fn = pool.getFn((v as TSerializedFn).__fn__) as {
      (__ctx__: unknown): unknown
      __deserialized: boolean
    }
    fn.__deserialized = true
    return fn as T
  }
  if (typeof v === 'object' && asObj) {
    const o = {} as Record<string, T>
    for (const [key, val] of Object.entries(v as Record<string, TSerializedFn | T>)) {
      o[key] = deserializeComputedFn(val) as T
    }
    return o as T
  }
  return v as T
}
