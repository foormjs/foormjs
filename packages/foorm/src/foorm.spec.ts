/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { deserializeForm } from './deserialize'
import { Foorm } from './foorm'
import { serializeForm } from './serialize'

interface TData {
  input: string
}
interface TContext {
  val: number
}

const foorm = new Foorm<TData, TContext>({
  title: () => 'My Title',
  entries: [
    {
      field: 'input',
      classes: 'class-a class-b',
      styles: 'max-width: 255px',
      label: 'Input Label',
      disabled: false,
      optional: true,
      validators: [(v: string) => v.length > 10, (v?: string) => Number(v?.length) > 10],
    },
    {
      field: 'field-2',
      type: 'type-2',
      classes: (_, _d, ctx) => `class-a class-b ${ctx.val}`,
      styles: { 'max-width': () => '255px', 'min-height': '100px' },
      label: (v: string) => `Field Label${v}`,
      disabled: (v, data) => !!data.input,
      optional: (v, data, ctx) => data.input.length < ctx.val,
    },
  ],
  submit: {
    text: () => 'Submit',
  },
})

describe('foorm', () => {
  it('must serialize into transferable object', () => {
    expect(serializeForm(foorm)).toMatchSnapshot()
  })
  it('must deserialize form', () => {
    const s = serializeForm(foorm)
    const newForm = deserializeForm(s)
    const def = newForm.getDefinition()
    expect((def.title as Function)({ v: '', data: {}, context: {} })).toEqual('My Title')
  })
})
it('must prepare validator', () => {
  foorm.setContext({ val: 2 })
  expect(
    foorm.getFormValidator()({
      'input': 't',
      'field-2': '',
    })
  ).toMatchInlineSnapshot(`
  {
    "errors": {
      "input": "Wrong value",
    },
    "passed": false,
  }
  `)
  expect(
    foorm.getFormValidator()({
      'input': 'test 123456789',
      'field-2': '',
    })
  ).toMatchInlineSnapshot(`
  {
    "errors": {
      "field-2": "Required",
    },
    "passed": false,
  }
  `)
  //   })
})
