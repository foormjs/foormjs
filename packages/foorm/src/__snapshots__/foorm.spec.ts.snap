// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foorm must serialize into transferable object 1`] = `
{
  "context": {},
  "entries": [
    {
      "attrs": undefined,
      "classes": "class-a class-b",
      "description": undefined,
      "disabled": false,
      "field": "input",
      "hidden": undefined,
      "hint": undefined,
      "label": "Input Label",
      "optional": true,
      "options": undefined,
      "placeholder": undefined,
      "styles": "max-width: 255px",
      "validators": [
        {
          "__fn__": "return ((v) => v.length > 10)(v,data,context,entry)",
        },
        {
          "__fn__": "return ((v) => Number(v === null || v === void 0 ? void 0 : v.length) > 10)(v,data,context,entry)",
        },
      ],
      "value": undefined,
    },
    {
      "attrs": undefined,
      "classes": {
        "__fn__": "return ((_, _d, ctx) => \`class-a class-b \${ctx.val}\`)(v,data,context,entry)",
      },
      "description": undefined,
      "disabled": {
        "__fn__": "return ((v, data) => !!data.input)(v,data,context,entry)",
      },
      "field": "field-2",
      "hidden": undefined,
      "hint": undefined,
      "label": {
        "__fn__": "return ((v) => \`Field Label\${v}\`)(v,data,context,entry)",
      },
      "optional": {
        "__fn__": "return ((v, data, ctx) => data.input.length < ctx.val)(v,data,context,entry)",
      },
      "options": undefined,
      "placeholder": undefined,
      "styles": {
        "max-width": {
          "__fn__": "return (() => '255px')(v,data,context,entry)",
        },
        "min-height": "100px",
      },
      "type": "type-2",
      "validators": [],
      "value": undefined,
    },
  ],
  "submit": {
    "disabled": undefined,
    "text": {
      "__fn__": "return (() => 'Submit')(data,context)",
    },
  },
  "title": {
    "__fn__": "return (() => 'My Title')(data,context)",
  },
}
`;
